### PR TITLE
Exclude empty outputs in requests

### PR DESCRIPTION
**Pull request by rjnrohit**
## Changes

Implements changes from these spec PRs:

* https://github.com/ethereum/EIPs/pull/8989
* https://github.com/ethereum/execution-apis/pull/599

## Types of changes
1. Exclude the empty requests from hash calculation
2. Add type prefix byte to the `ExecutionRequests` for the engineAPIs
#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [ ] Yes
- [x] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/ExecutionRequests/ExecutionRequestProcessor.cs [hunk 1]
```diff
@@ -54,12 +54,15 @@ public ExecutionRequestsProcessor(ITransactionProcessor transactionProcessor)
         _consolidationTransaction.Hash = _consolidationTransaction.CalculateHash();
     }
 
-    public byte[] ProcessDeposits(TxReceipt[] receipts, IReleaseSpec spec)
+    public void ProcessDeposits(TxReceipt[] receipts, IReleaseSpec spec, ref ArrayPoolList<byte[]> requests)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/ExecutionRequests/ExecutionRequestProcessor.cs:57]: `ref` not needed it is reference type

File: src/Nethermind/Nethermind.Consensus/ExecutionRequests/ExecutionRequestProcessor.cs [hunk 2]
```diff
@@ -107,34 +111,51 @@ private void DecodeDepositRequest(LogEntry log, Span<byte> buffer)
...
+            ProcessDeposits(receipts, spec, ref requests);
+            ReadRequests(block, state, spec, spec.Eip7002ContractAddress, ref requests);
+            ReadRequests(block, state, spec, spec.Eip7251ContractAddress, ref requests);
+            block.ExecutionRequests = requests.ToArray();
+            block.Header.RequestsHash = ExecutionRequestExtensions.CalculateHashFromFlatEncodedRequests(block.ExecutionRequests);
+        }
+        finally
+        {
+            requests.Dispose();
+        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/ExecutionRequests/ExecutionRequestProcessor.cs:159]:
```suggestion
        using ArrayPoolList<byte[]> requests = new(3);
            ProcessDeposits(receipts, spec, ref requests);
            ReadRequests(block, state, spec, spec.Eip7002ContractAddress, ref requests);
            ReadRequests(block, state, spec, spec.Eip7251ContractAddress, ref requests);
            block.ExecutionRequests = requests.ToArray();
            block.Header.RequestsHash = ExecutionRequestExtensions.CalculateHashFromFlatEncodedRequests(block.ExecutionRequests);
```

File: src/Nethermind/Nethermind.Core/ExecutionRequest/ExecutionRequestExtensions.cs [hunk 3]
```diff
@@ -38,25 +38,17 @@ public static int GetRequestsByteSize(this IEnumerable<ExecutionRequest> request
...
+        using ArrayPoolList<byte> concatenatedHashes = new(Hash256.Size * MaxRequestsCount);
         foreach (byte[] requests in flatEncodedRequests)
         {
-            requestBuffer[0] = type;
-            requests.CopyTo(requestBuffer.Slice(1, requests.Length));
-            sha256.ComputeHash(requestBuffer.Slice(0, requests.Length + 1).ToArray()).CopyTo(concatenatedHashes.AsSpan(currentPosition, Hash256.Size));
-            currentPosition += Hash256.Size;
-            type++;
+            if (requests.Length <= 1) continue;
+            concatenatedHashes.AddRange(sha256.ComputeHash(requests));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/ExecutionRequest/ExecutionRequestExtensions.cs:51]: Why `AddRange` instead of `Add`?

Comment (rjnrohit) [reply src/Nethermind/Nethermind.Core/ExecutionRequest/ExecutionRequestExtensions.cs:51]: `sha256.ComputeHash(requests)` returns an array of byte

**Review by MarekM25 — APPROVED** [review]
Let's hold off on merging this for now as it may break Mekong. The changes look good overall.

**Pull request merged**